From 7a1feb1efcbe1a803db47bffbbafea1ddd730518 Mon Sep 17 00:00:00 2001
From: Alejandro Hernandez <alejandro.hernandez@linux.intel.com>
Date: Mon, 6 Jul 2015 10:17:19 +0000
Subject: [PATCH] opkg-build: Ignore tar error due to hardlinks issue when

If a the number of hard links decreases or increases while creating
the tar files used for an ipk package, tar fails with error code 1:

| DEBUG: Executing python function do_package_ipk
| tar: ./usr/src/debug/gperf/3.0.4-r0/gperf-3.0.4/src/main.cc: file changed as we read it
NOTE: recipe gperf-3.0.4-r0: task do_package_write_ipk: Failed
ERROR: Task 6539 (recipes-extended/gperf/gperf_3.0.4.bb, do_package_write_ipk) failed with exit code '1'

We detect if the error code produced by tar is 1 and in this case ignore it.

This a similar behavior to the one on dpkg:
http://git.yoctoproject.org/cgit/cgit.cgi/poky/commit/?id=40731942515ec8d80c727ad561174986d4f05818

Upstream-Status: Inappropriate

Signed-off-by: Alejandro Hernandez <alejandro.hernandez@linux.intel.com>

---
 opkg-build | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/opkg-build b/opkg-build
index 532b6dc..234182e 100755
--- a/opkg-build
+++ b/opkg-build
@@ -291,8 +291,21 @@ mkdir $tmp_dir
 build_date="$(date --date="@${SOURCE_DATE_EPOCH:-$(date +%s)}" "+%Y-%m-%d %H:%M:%S")"
 
 echo $CONTROL > $tmp_dir/tarX
-( cd $pkg_dir && tar $ogargs $tsortargs -X $tmp_dir/tarX -c $tarformat . | $compressor $compressorargs > $tmp_dir/data.tar.$cext )
-( cd $pkg_dir/$CONTROL && tar $ogargs $tsortargs --mtime="$build_date" -c $tarformat . | gzip $zipargs > $tmp_dir/control.tar.gz )
+
+
+# Ignore error code 1, caused by modifying the number of hard links while creating the tar file
+rc=0
+( cd $pkg_dir && tar $ogargs $tsortargs -X $tmp_dir/tarX -c $tarformat . | $compressor $compressorargs > $tmp_dir/data.tar.$cext ) || rc=$?
+if [ $rc -ne 1 ] && [ $rc -ne 0 ]; then
+        exit $rc 
+fi
+
+rc=0
+( cd $pkg_dir/$CONTROL && tar $ogargs $tsortargs --mtime="$build_date" -c $tarformat . | gzip $zipargs > $tmp_dir/control.tar.gz ) || rc=$?
+if [ $rc -ne 1 ] && [ $rc -ne 0 ]; then
+        exit $rc
+fi
+
 rm $tmp_dir/tarX
 
 echo "2.0" > $tmp_dir/debian-binary
